crates/goose-server/src/tunnel/mod.rs

```diff
@@ -0,0 +1,263 @@
+            }
+            None => TunnelInfo {
+                state,
+                url: None,
+                hostname: None,
+                secret: None,
+            },
+        }
+    }
+
+    pub async fn update_config<F>(&self, f: F)
+    where
+        F: FnOnce(&mut TunnelConfig),
+    {
+        let mut cfg = self.config.write().await;
+        f(&mut cfg);
+        if let Err(e) = config::save_config(&cfg).await {
+            tracing::error!("Failed to save tunnel config: {}", e);
+        }
+    }
+
+    pub async fn start(&self) -> anyhow::Result<TunnelInfo> {
+        let mut state = self.state.write().await;
+        if *state != TunnelState::Idle {
+            anyhow::bail!("Tunnel is already running or starting");
+        }
+        *state = TunnelState::Starting;
+        drop(state);
+
+        let config = self.config.read().await.clone();
+        let server_port = get_server_port()?;
+
+        let tunnel_secret = config.secret.clone().unwrap_or_else(generate_secret);
+        let server_secret =
+            std::env::var("GOOSE_SERVER__SECRET_KEY").unwrap_or_else(|_| "test".to_string());
+        let agent_id = config.agent_id.clone().unwrap_or_else(generate_agent_id);
+
+        self.update_config(|c| {
+            c.secret = Some(tunnel_secret.clone());
+            c.agent_id = Some(agent_id.clone());
+        })
+        .await;
+
+        let (restart_tx, mut restart_rx) = mpsc::channel::<()>(1);
+        *self.restart_tx.write().await = Some(restart_tx.clone());
+
+        let result = lapstone::start(
+            server_port,
+            tunnel_secret,
+            server_secret,
+            agent_id,
+            self.lapstone_handle.clone(),
+            restart_tx,
+        )
+        .await;
+
+        match result {
+            Ok(info) => {
+                *self.state.write().await = TunnelState::Running;
+                *self.info.write().await = Some(info.clone());
+                self.update_config(|c| c.auto_start = true).await;
+
+                let state = self.state.clone();
+                let config = self.config.clone();
+                let lapstone_handle = self.lapstone_handle.clone();
+                let restart_tx_clone = self.restart_tx.clone();
+                let watchdog_handle_arc = self.watchdog_handle.clone();
+
+                let watchdog = tokio::spawn(async move {
+                    while restart_rx.recv().await.is_some() {
+                        let auto_start = config.read().await.auto_start;
+                        if auto_start {
+                            tracing::warn!("Tunnel connection lost, initiating restart...");
+
+                            lapstone::stop(lapstone_handle.clone()).await;
+                            *state.write().await = TunnelState::Idle;
+
+                            tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+
+                            *state.write().await = TunnelState::Starting;
```

Comment:
the restart logic here is rather similar to what we are doing with the starting logic. I would suggest factoring it out in one method that you call twice

Reply:
yes good idea - the restart is a fairly late addition (where I made it actually do similar to start - in fact, I want it to do EXACTLY the same thing, so yes).